### PR TITLE
feat!: support for local repos

### PR DESCRIPTION
**Pull request by D-Brox**
## Purpose

Add support for locally hosted repos

## Approach

Use `curl` instead of `wget` due to it's  `file://` protocol support. Format path  in `add-repo` when adding local repos.

## Checklist

- [x] I confirm that I have read the [contributing guidelines](https://github.com/pacstall/pacstall/blob/develop/CONTRIBUTING.md), and this pull request is abiding by all the clauses stated in the guideline.

### CODE REVIEW

**Review by Henryws — CHANGES_REQUESTED** [review]
If a secondary repo is misconfigured, operations like install and download fail.

Make sure to allow properly configured repos to work even if a misconfigured one exists. Add a warning about the misconfigured one

File: pacstall [hunk 1]
```diff
@@ -194,25 +194,35 @@ function ask() {
...
-			fancy_message error "Failed with http code ${http_code}"
-			return 1
-			;;
-	esac
+    if [[ "${1}" == "file://"* ]]; then
+    	if [[ -f "${1}" ]]; then
+    	  true
+    	fi
+    else
+		http_code="$(curl -o /dev/null -s --head --write-out '%{http_code}\n' -- "${1}")"
```
Comment (Henryws) [inline pacstall:202]: This doesn't have the same tabbing as lines above

**Review by Henryws — CHANGES_REQUESTED** [review]

File: pacstall (hunk 1 above)
Comment (D-Brox) [reply pacstall:202]: What do you mean?

**Review by Henryws — CHANGES_REQUESTED** [review]
image

File: misc/scripts/search.sh [hunk 2]
```diff
@@ -73,8 +90,16 @@ fi
 PACKAGELIST=()
 URLLIST=()
 while IFS= read -r URL; do
+	if [[ ${URL} == "/"* ]] || [[ ${URL} == "~"* ]] || [[ ${URL} == "."* ]]; then
+		sed -i "s#${URL}#file://$(readlink -f ${URL})#g" "$STGDIR/repo/pacstallrepo.txt" ||
+		fancy_message warn "Add `file://` to the local repo absolute path on $CYAN$STGDIR/repo/pacstallrepo.txt$NC`
```
Comment (Henryws) [inline misc/scripts/search.sh:95]: This has a dangling quote "

**Review by Henryws — CHANGES_REQUESTED** [review]

File: misc/scripts/search.sh (hunk 2 above)
Comment (D-Brox) [reply misc/scripts/search.sh]: Muscle memory strikes again

**Review by Henryws — CHANGES_REQUESTED** [review]
image
The showed text for local repos should be something like `/home/henry/local-repo-pacstall-test`, instead of having the `file://` in front of it

**Review by Henryws — CHANGES_REQUESTED** [review]
image
This `@` should not be the same color as the repo

File: misc/scripts/download.sh [hunk 3]
```diff
@@ -35,12 +35,14 @@ if check_url "${URL}"; then
 	fi
 
 	case "$URL" in
-		*.pacscript)
-			if ! wget -q --show-progress --progress=bar:force -- "$URL" > /dev/null 2>&1; then
+		*.pacscript|*packagelist)
+			FILE=$(basename "$URL")
+			if ! curl -s -- "$URL" > $FILE; then
```
Comment (Henryws) [inline misc/scripts/download.sh:40]:
```suggestion
			if ! curl -s -- "$URL" > "$FILE"; then
```

**Review by Henryws — CHANGES_REQUESTED** [review]

**Review by Henryws — CHANGES_REQUESTED** [review]
image
Yet the repo does exist. This error also happens with URLs as well

File: misc/scripts/search.sh [hunk 4]
```diff
@@ -28,21 +28,55 @@ if [[ -n $UPGRADE ]]; then
...
+		local REPODIR="$(getPath ${REPO})"
+		echo "\e]8;;$REPO\a$REPODIR\e]8;;\a"
+	else
+		echo "\e]8;;$REPO\a$REPO\e]8;;\a"
+	fi
+}
+
 if [[ $PACKAGE == *@* ]]; then
 	REPONAME=${PACKAGE#*@}
+	if [[ $REPONAME == "file://" ]] ||[[ $REPONAME == "/"* ]] || [[ $REPONAME == "~"* ]] || [[ $REPONAME == "."* ]]; then
```
Comment (Henryws) [inline misc/scripts/search.sh:77]:
```suggestion
	if [[ $REPONAME == "file://" ]] || [[ $REPONAME == "/"* ]] || [[ $REPONAME == "~"* ]] || [[ $REPONAME == "."* ]]; then
```

**Review by Henryws — CHANGES_REQUESTED** [review]

**Review by Henryws — APPROVED** [review]

File: misc/scripts/add-repo.sh [hunk 5]
```diff
@@ -49,7 +53,7 @@ else
 	fi
 fi
 
-if ! wget -q --spider -- "$REPO/packagelist"; then
+if ! curl --head -s --fail -- "$REPO/packagelist" > /dev/null; then
```
Comment (wizard-28) [inline misc/scripts/add-repo.sh:56]:
```suggestion
if ! curl --head --location -s --fail -- "$REPO/packagelist" > /dev/null; then
```

Makes it follow redirects.

File: misc/scripts/download.sh [hunk 6]
```diff
@@ -35,12 +35,14 @@ if check_url "${URL}"; then
 	fi
 
 	case "$URL" in
-		*.pacscript)
-			if ! wget -q --show-progress --progress=bar:force -- "$URL" > /dev/null 2>&1; then
+		*.pacscript | *packagelist)
+			FILE=$(basename "$URL")
+			if ! curl -s -- "$URL" > "$FILE"; then
```
Comment (wizard-28) [inline misc/scripts/download.sh:40]:
```suggestion
			if ! curl --location -s -- "$URL" > "$FILE"; then
```

**Review by wizard-28 — CHANGES_REQUESTED** [review]

File: pacstall [hunk 7]
```diff
@@ -194,25 +194,35 @@ function ask() {
...
 			true
-			;;
-		*)
-			fancy_message error "Failed with http code ${http_code}"
-			return 1
-			;;
-	esac
+		fi
+	else
+		http_code="$(curl -o /dev/null -s --head --write-out '%{http_code}\n' -- "${1}")"
```
Comment (wizard-28) [inline pacstall:202]:
```suggestion
		http_code="$(curl --location -o /dev/null -s --head --write-out '%{http_code}\n' -- "${1}")"
```

Shouldn't this have it too?

**Review by wizard-28 — CHANGES_REQUESTED** [review]

**Review by wizard-28 — CHANGES_REQUESTED** [review]
If the repository is deleted without prior removal from the repolist, then it should warn the user about it, in all commands related to the package repos

It doesn't do that currently

**Review by wizard-28 — APPROVED** [review]

**Pull request merged**